test/server/listener_manager_impl_test.cc

```diff
@@ -213,6 +213,79 @@ class ListenerManagerImplWithRealFiltersTest : public ListenerManagerImplTest {
     return result;
   }
 
+  /**
+   * Create an IPv4 listener with a given name.
+   */
+  envoy::api::v2::Listener createIPv4Listener(const std::string& name) {
+    envoy::api::v2::Listener listener = parseListenerFromV2Yaml(R"EOF(
+      address:
+        socket_address: { address: 127.0.0.1, port_value: 1111 }
+      filter_chains:
+      - filters:
+    )EOF");
+    listener.set_name(name);
+    return listener;
+  }
+
+  /**
+   * Validate that createListenSocket is called once with the expected options.
+   */
+  void
+  expectCreateListenSocket(const envoy::api::v2::core::SocketOption::SocketState& expected_state,
+                           Network::Socket::Options::size_type expected_num_options) {
+    EXPECT_CALL(listener_factory_, createListenSocket(_, _, _, true))
+        .WillOnce(Invoke([this, expected_num_options, &expected_state](
+                             Network::Address::InstanceConstSharedPtr, Network::Address::SocketType,
+                             const Network::Socket::OptionsSharedPtr& options,
+                             bool) -> Network::SocketSharedPtr {
+          EXPECT_NE(options.get(), nullptr);
+          EXPECT_EQ(options->size(), expected_num_options);
+          EXPECT_TRUE(
+              Network::Socket::applyOptions(options, *listener_factory_.socket_, expected_state));
+          return listener_factory_.socket_;
+        }));
+  }
+
+  /**
+   * Validate that setsockopt() is called the expected number of times with the expected options.
+   */
+  void expectSetsockopt(NiceMock<Api::MockOsSysCalls>& os_sys_calls, int expected_sockopt_level,
+                        int expected_sockopt_name, int expected_value,
+                        uint32_t expected_num_calls = 1) {
+    EXPECT_CALL(os_sys_calls,
+                setsockopt_(_, expected_sockopt_level, expected_sockopt_name, _, sizeof(int)))
+        .Times(expected_num_calls)
+        .WillRepeatedly(
+            Invoke([expected_value](int, int, int, const void* optval, socklen_t) -> int {
+              EXPECT_EQ(expected_value, *static_cast<const int*>(optval));
+              return 0;
+            }));
+  }
+
+  /**
```

Comment:
Nit: we tend to use `//` for internal implementation comments, and Doxygen style for API level comments. But we're not 100% consistent, so just an FYI for next PRs.